src/gc-pages.c

```diff
@@ -56,231 +55,59 @@ static char *jl_gc_try_alloc_pages(int pg_cnt) JL_NOTSAFEPOINT
-                        }
-                    }
-                    info.pagetable1->freemap0[info.pagetable1_i32] &= ~(uint32_t)(1u << info.pagetable1_i); // record that this was full
-                }
-            }
-            memory_map.freemap1[info.pagetable_i32] &= ~(uint32_t)(1u << info.pagetable_i); // record that this was full
-        }
-    }
+    jl_gc_pagemeta_t *meta = NULL;
+    jl_mutex_lock_nogc(&global_page_pool.lock);
 
-    // no existing pages found, allocate a new one
-    {
-        jl_gc_pagemeta_t *meta = jl_gc_alloc_new_page();
-        info = page_metadata_ext(meta->data);
-        assert(meta == info.meta);
-        // new pages are now available starting at max of lb and pagetable_i32
-        if (memory_map.lb > info.pagetable_i32)
-            memory_map.lb = info.pagetable_i32;
-        if (info.pagetable1->lb > info.pagetable1_i32)
-            info.pagetable1->lb = info.pagetable1_i32;
-        if (info.pagetable0->lb > info.pagetable0_i32)
-            info.pagetable0->lb = info.pagetable0_i32;
+    if (global_page_pool.page_metadata_back != NULL) {
+        meta = pop_page_metadata_back(&global_page_pool.page_metadata_back);
+        jl_mutex_unlock_nogc(&global_page_pool.lock);
+        gc_alloc_map_set(meta->data, 1);
     }
 
-have_free_page:
-    // in-use pages are now ending at min of ub and pagetable_i32
-    if (memory_map.ub < info.pagetable_i32)
-        memory_map.ub = info.pagetable_i32;
-    if (info.pagetable1->ub < info.pagetable1_i32)
-        info.pagetable1->ub = info.pagetable1_i32;
-    if (info.pagetable0->ub < info.pagetable0_i32)
-        info.pagetable0->ub = info.pagetable0_i32;
-
-    // mark this entry as in-use and not free
-    info.pagetable0->freemap[info.pagetable0_i32] &= ~(uint32_t)(1u << info.pagetable0_i);
-    info.pagetable0->allocmap[info.pagetable0_i32] |= (uint32_t)(1u << info.pagetable0_i);
-    info.pagetable1->allocmap0[info.pagetable1_i32] |= (uint32_t)(1u << info.pagetable1_i);
-    memory_map.allocmap1[info.pagetable_i32] |= (uint32_t)(1u << info.pagetable_i);
-
-#ifdef _OS_WINDOWS_
-    VirtualAlloc(info.meta->data, GC_PAGE_SZ, MEM_COMMIT, PAGE_READWRITE);
-#endif
+    if (meta == NULL) {
+        jl_mutex_unlock_nogc(&global_page_pool.lock);
+        char *data = jl_gc_try_alloc_pages();
+        if (data == NULL) {
+            jl_throw(jl_memory_exception);
+        }
+        meta = (jl_gc_pagemeta_t*)malloc_s(block_pg_cnt * sizeof(jl_gc_pagemeta_t));
+        for (int i = 0; i < block_pg_cnt; i++) {
+            jl_gc_pagemeta_t *pg = &meta[i];
+            pg->data = data + GC_PAGE_SZ * i;
+            jl_mutex_lock_nogc(&alloc_map.lock);
+            gc_alloc_map_maybe_create(pg->data);
+            jl_mutex_unlock_nogc(&alloc_map.lock);
+            if (i == 0) {
+                gc_alloc_map_set(pg->data, 1);
+            }
+            else {
+                jl_mutex_lock_nogc(&global_page_pool.lock);
```

Comment:
Seems awkward that we need to acquire this lock over 10k times to allocate this metadata. The layout of the `global_page_pool.page_metadata_back` list seems like it also might end up quite poor by the end of that.